### PR TITLE
ADAPTIVE_FAN_SLOWING extension to hotend thermal protection

### PR DESCRIPTION
**Pull request by InsanityAutomation**
Partial menu restructure for fans to use helper functions to allow function to be parsed properly. Reincorporate adaptive fan slowing in a different way to get some math out of the temp isr.

Needs a few more tweaks and a bit more testing but should be closer to where we need to go.

### CODE REVIEW

**Comment by InsanityAutomation** [comment]
@marcio-ao Should i scale the fan speed output to extui if triggered or keep the setpoint sent?

**Comment by InsanityAutomation** [comment]
Working good and updating the screens now. Just need to do a little work on the menu display and this should be ready... Thats a job for tomorrow. May add a note to the config that a larger runaway hysteresis gives this a wider degree of control and limits pulsating fan speeds.

**Comment by marcio-ao** [comment]
@InsanityAutomation: I'm not sure I understand what adaptive fan slowing is for or how it works, but it looks like you already updated the relevant UI methods.

**Comment by InsanityAutomation** [comment]
@marcio-ao with your Aerostruder if I set the fan to 100% the hotend can't keep up in most cases and causes thermal runaway. This throttles the fan speed back if temp is falling to prevent it if possible. Given this, actual output speed may be less than fan_speed[] so I was scaling the display and blinking between setpoint and actual.

**Comment by marcio-ao** [comment]
@InsanityAutomation: I think I understand your question now. It looks like ADAPTIVE_FAN_SCALING will lower the fan speed if the temperature isn't keeping up. Your question probably related to whether in the ExtUI, the "getFan_percent" method should be reduced. I would say no, for the following reason: the UI shows the value of "getFan_percent" in the edit boxes, it then calls "setFan_percent" with that same value plus an increment. So for the controls to work properly, "setFan_percent" and "getFan_percent" must be the *requested* fan speed, not the actual fan speed.

So your PR brings up an interesting question. If the actual fan speed may not be the requested fan speed, then we need different methods to make that distinction clear. I would think the way to do it would be to mirror the temperature methods, it would look like this:

- getTargetFan_percent (presently getFan_percent)
- setTargetFan_percent (presently setFan_percent)
- getActualFan_percent

That way, get/setTargetFan would be used for the edit boxes, while on the status screen the UI designer could choose to display either the target or the actual fan speed (perhaps both, like temps).

If you want to add these changes to the API, can I can review them and make sure they look okay.

**Comment by InsanityAutomation** [comment]
Ok, I'll poke at that after I fix the graphical LCD menu items to work with this.

**Comment by InsanityAutomation** [comment]
Ok I glanced and it was simple so I threw it in quick.

**Comment by marcio-ao** [comment]
@InsanityAutomation: Yup, that's the idea. I think that will cover all cases. Thank you!

File: Marlin/src/lcd/extensible_ui/ui_api.cpp [hunk 1]
```diff
@@ -180,7 +180,9 @@ namespace ExtUI {
     return thermalManager.degTargetHotend(extruder - E0);
   }
 
-  float getFan_percent(const fan_t fan) { return ((float(fan_speed[fan - FAN0]) + 1) * 100) / 256; }
+  float getTargetFan_percent(const fan_t fan) { return ((float(fan_speed[fan - FAN0]) + 1) * 100) / 256; }
+
+  float getActualFan_percent(const fan_t fan) { return ((int(fan_speed[0] * (fan_speed_multiplier[0]/100.0f)) + 1) * 100) / 256; }
```
Comment (marcio-ao) [inline Marlin/src/lcd/extensible_ui/ui_api.cpp]: This is only operating on fan zero, however.

Comment (InsanityAutomation) [reply Marlin/src/lcd/extensible_ui/ui_api.cpp]: Fixed :)

**Comment by InsanityAutomation** [comment]
That should do it.

BTW, any idea if you guys will they be showing off a 7 at RAPID this year? Since its only about 30min from my house this year, I have no excuse to miss it!

**Comment by marcio-ao** [comment]
> That should do it.

Looks good now, thanks!

> BTW, any idea if you guys will they be showing off a 7 at RAPID this year? Since its only about 30min from my house this year, I have no excuse to miss it!

I asked and all I got was that it likely we will be at RAPID, it is too early to know for sure what will be demonstrated. You should look for our booth anyway, I'm sure it will be something cool :)

**Comment by InsanityAutomation** [comment]
Yeah, I saw all the bungee cord machines at IMTS last year. That was kinda fun but carefully set so Y wasn't going vertically :)

**Comment by marcio-ao** [comment]
> Yeah, I saw all the bungee cord machines at IMTS last year. That was kinda fun but carefully set so Y wasn't going vertically :)

Shhhhh... :)

BTW, I think we need a #ifdef inside  `getActualFan_percent`. When ADAPTIVE_FAN_SLOWING is DISABLED, it probably ought to just call "getTargetFan_percent` and return that.

**Comment by InsanityAutomation** [comment]
In part because fan_speed_multiplier wouldnt be defined without it... so yeah kinda need that :)

**Comment by thinkyhead** [comment]
To get the contents of a rebased / squashed branch, you must:
```
git fetch origin
git checkout FanLCDRefactor
git reset --hard origin/FanLCDRefactor
```

**Comment by InsanityAutomation** [comment]
Yeah, I hit the wrong button and just accepted all the incoming.... Then I just rolled with it since at that point it was just a matter of a nuisance merge branch line

**Comment by thinkyhead** [comment]
No problem. `git diff` comes in handy for this kind of thing.

**Comment by InsanityAutomation** [comment]
Youre a faster git-ninja than I am lol

**Pull request merged**